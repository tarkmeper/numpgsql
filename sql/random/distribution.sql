SELECT random_seed(1);

SELECT random_beta(2,3, '{2,3}'::integer[]);
SELECT random_beta(-1,3, '{2,3}'::integer[]);
SELECT random_beta(2,-3, '{2,3}'::integer[]);

SELECT random_binomial(2,0.5, '{2,3}'::integer[]);
SELECT random_binomial(-2,0.5, '{2,3}'::integer[]);
SELECT random_binomial(2,3, '{2,3}'::integer[]);
SELECT random_binomial(2,-3, '{2,3}'::integer[]);



