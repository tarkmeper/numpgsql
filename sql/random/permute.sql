SELECT random_seed(1);

SELECT random_permute('{}'::integer[]);
SELECT random_permute('{1,2,3,4,5,6}'::integer[]);
SELECT random_permute('{{1,2,3},{4,5,6}}'::integer[]);

-- Choice without replacement
SELECT random_choice('{}'::integer[], '{3}'::integer[], false);
SELECT random_choice('{1,2,3,4,5}'::integer[], '{3}'::integer[], false);
SELECT random_choice('{1,2,3,4,5}'::integer[], '{2,2}'::integer[], false);
SELECT random_choice('{1,2,3,4,5}'::integer[], '{7}'::integer[], false);
SELECT random_choice('{1,2,3,4,5}'::integer[], '{3,2}'::integer[], false);


-- Choice with replacement
SELECT random_choice('{}'::integer[], '{3}'::integer[], true);
SELECT random_choice('{1,2,3,4,5}'::integer[], '{3}'::integer[], true);
SELECT random_choice('{1,2,3,4,5}'::integer[], '{2,2}'::integer[], true);
SELECT random_choice('{1,2,3,4,5}'::integer[], '{7}'::integer[], true);
SELECT random_choice('{1,2,3,4,5}'::integer[], '{3,2}'::integer[], true);
SELECT random_choice('{1,2,3,4,5}'::integer[], '{3,2,4,2,1,2,3}'::integer[], true);
SELECT random_choice('{1,2,3,4,5}'::integer[], '{{3,2,1,4},{2,1,2,3}}'::integer[], true);
