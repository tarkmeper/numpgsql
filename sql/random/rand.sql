SELECT random_seed(1);

SELECT random_rand(2);

SELECT random_rand(2,2);

SELECT random_rand();

SELECT random_rand(2,2,2,2,2,2,2,2);



