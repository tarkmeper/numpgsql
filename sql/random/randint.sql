SELECT random_seed(1);

SELECT random_randint(1,5,'{2}'::integer[]);

SELECT random_randint(1,5,'{2,2}'::integer[]);

-- some error-ish conditions
SELECT random_randint(5,1,'{2,2}'::integer[]);
SELECT random_randint(1,1,'{2,2}'::integer[]);
