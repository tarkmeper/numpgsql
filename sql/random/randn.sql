SELECT random_seed(1);

SELECT random_randn(2);

SELECT random_randn(2,2);




