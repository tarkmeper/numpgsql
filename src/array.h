extern "C"  {
        #include "postgres.h"
        #include "catalog/pg_type.h"
        #include "utils/array.h"
        #include "fmgr.h"
}
#include <vector>
#include <tuple>


inline static size_t arr_cnt(const ArrayType* arr) {
	size_t result = 1;
	for (int i = 0; i < ARR_NDIM(arr); ++i){
		result *= ARR_DIMS(arr)[i];
	}
	return result;
}

//Return values - TODO rename function this was poorly named.
template<class T> inline Datum _ret(const T& val) { T::unimplemented_function; return Datum(); }
template<> inline Datum _ret<double>(const double& val) { PG_RETURN_FLOAT8(val); }
template<> inline Datum _ret<float>(const float& val) { PG_RETURN_FLOAT4(val); }
template<> inline Datum _ret<short>(const short& val) { PG_RETURN_INT16(val); }
template<> inline Datum _ret<int>(const int& val) { PG_RETURN_INT32(val); }
template<> inline Datum _ret<long>(const long& val) { PG_RETURN_INT64(val); }
template<> inline Datum _ret<bool>(const bool& val) { PG_RETURN_BOOL(val); }

//OID types
template<class T> inline Oid oid_type() { return T::unimplemented_feature; }
template<> inline Oid oid_type<float>() { return FLOAT4OID; }
template<> inline Oid oid_type<double>() { return FLOAT8OID; }
template<> inline Oid oid_type<short>() { return INT2OID; }
template<> inline Oid oid_type<int>() { return INT4OID; }
template<> inline Oid oid_type<long>() { return INT8OID; }
template<> inline Oid oid_type<bool>() { return BOOLOID; }

//Alignment codes
template<class T> char oid_align() { return T::unimplemented_feature; }
template<> inline char oid_align<float>() { return 'i'; }
template<> inline char oid_align<double>() { return 'd'; }
template<> inline char oid_align<short>() { return 's'; }
template<> inline char oid_align<int>() { return 'i'; }
template<> inline char oid_align<long>() { return 'd'; }
template<> inline char oid_align<bool>() { return 'c'; }

//Get param values
template<class T> inline T get_param(PG_FUNCTION_ARGS, int p)  { T::unimplemented_function; }
template<> inline bool get_param<bool>(PG_FUNCTION_ARGS, int p) { return PG_GETARG_BOOL(p); }
template<> inline float get_param<float>(PG_FUNCTION_ARGS, int p) { return PG_GETARG_FLOAT4(p); }
template<> inline double get_param<double>(PG_FUNCTION_ARGS, int p) { return PG_GETARG_FLOAT8(p); }
template<> inline short get_param<short>(PG_FUNCTION_ARGS, int p) { return PG_GETARG_INT16(p); }
template<> inline int get_param<int>(PG_FUNCTION_ARGS, int p) { return PG_GETARG_INT32(p); }
template<> inline long get_param<long>(PG_FUNCTION_ARGS, int p) { return PG_GETARG_INT64(p); }
