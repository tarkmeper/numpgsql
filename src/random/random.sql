-- This program is open source, licensed under the MIT License.
-- For license terms, see the LICENSE file.
--
-- Copyright (C) 2019: Mark Teper

-- Other functions

CREATE FUNCTION random_seed(integer) RETURNS void LANGUAGE C STRICT COST 100 AS '$libdir/numpgsql', 'set_seed';
CREATE FUNCTION random_state() RETURNS text LANGUAGE C STRICT COST 100 AS '$libdir/numpgsql', 'get_rgen_state';
CREATE FUNCTION random_state(text) RETURNS void LANGUAGE C STRICT COST 100 AS '$libdir/numpgsql', 'set_rgen_state';

CREATE FUNCTION random_rand(VARIADIC integer[]) RETURNS double precision[] LANGUAGE C STRICT COST 100 AS '$libdir/numpgsql', 'gen_rand';
CREATE FUNCTION random_randn(VARIADIC integer[]) RETURNS double precision[] LANGUAGE C STRICT COST 100 AS '$libdir/numpgsql', 'gen_randn';
CREATE FUNCTION random_randint(low bigint, high bigint, integer[]) RETURNS bigint[] LANGUAGE C STRICT COST 100 AS '$libdir/numpgsql', 'gen_randint';

CREATE FUNCTION random_permute(anyarray) RETURNS anyarray LANGUAGE C STRICT COST 100 AS '$libdir/numpgsql', 'permute';
CREATE FUNCTION random_choice(anyarray, shape int[], replace boolean) RETURNS anyarray LANGUAGE C STRICT COST 100 AS '$libdir/numpgsql', 'choice';


CREATE FUNCTION random_beta(alpha double precision, beta double precision, shape int[]) RETURNS double precision[] LANGUAGE C STRICT COST 100 AS '$libdir/numpgsql', 'gen_beta';
CREATE FUNCTION random_binomial(t int, p double precision, shape int[]) RETURNS integer[] LANGUAGE C STRICT COST 100 AS '$libdir/numpgsql', 'gen_binomial';

