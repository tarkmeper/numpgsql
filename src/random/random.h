#ifndef RANDOM_H
#define RANDOM_H

#include <random>

extern std::mt19937 rgen;

struct ArrDims {
   int ndims;
   int* dims;
   int lbs[MAXDIM];
   size_t n;
};

extern ArrDims extract_shape(ArrayType* arr);

#endif
