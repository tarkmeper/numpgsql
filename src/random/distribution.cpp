extern "C"  {
        #include "postgres.h"
        #include "catalog/pg_type.h"
        #include "fmgr.h"
}
#include "random.h"
#include "../array.h"

#include <boost/random.hpp>

template<class D> static Datum _gen_rand(PG_FUNCTION_ARGS, D& dist, int dim_arr_idx = 0) {
  typedef typename D::result_type T;
  ArrayType* arr = PG_GETARG_ARRAYTYPE_P_COPY(dim_arr_idx);

  if (ARR_NDIM(arr) == 0 ) {
     PG_RETURN_POINTER(construct_empty_array(oid_type<double>()));
  } else if (ARR_NDIM(arr) != 1 ) {
     elog(ERROR, "GEN_RAND: Size specification can only be 1-dimension array");
     PG_RETURN_NULL();
  } else if (ARR_DIMS(arr)[0] >= 4 ) {
     elog(ERROR, "GEN_RAND: Maximum of 4 dimensions in array generation.");
     PG_RETURN_NULL();
  }

  //compute number of dimensions
  ArrDims rd = extract_shape(arr);
  Datum* r = (Datum*)palloc(rd.n * sizeof(Datum));
  for (size_t i = 0; i < rd.n; ++i ) {
     r[i] = _ret<T>(dist(rgen));
  }

  ArrayType* res = construct_md_array(r,
                                      NULL,
                                      rd.ndims,
                                      rd.dims,
                                      rd.lbs,
                                      oid_type<T>(),
                                      sizeof(T),
                                      true,
                                      oid_align<T>());
  PG_RETURN_POINTER(res);
}

#define RAND_FUNCTION(NAME) \
	extern "C" { PG_FUNCTION_INFO_V1(NAME); Datum NAME(PG_FUNCTION_ARGS); } \
        Datum NAME(PG_FUNCTION_ARGS)

RAND_FUNCTION(gen_rand) {
  std::uniform_real_distribution<double> dist{0.0, 1.0};
  return _gen_rand(fcinfo, dist);
}


RAND_FUNCTION(gen_randn) {
  //for soem reason can't determine "std' variant is producing warnings.
  std::normal_distribution<double> dist{0.0, 1.0};
  return _gen_rand(fcinfo, dist);
}


RAND_FUNCTION(gen_randint) {
  long low = get_param<long>(fcinfo, 0);
  long high = get_param<long>(fcinfo, 1);
  if (high <= low ){
     elog(ERROR, "GEN_RANDINT: Cannot have high bound lower then low bound");
     PG_RETURN_NULL();
  } else {
     std::uniform_int_distribution<> dist(low, high);
     return _gen_rand(fcinfo, dist, 2);
  }
}

RAND_FUNCTION(gen_beta) {
  float alpha = get_param<double>(fcinfo, 0);
  float beta  = get_param<double>(fcinfo, 1);
  if (alpha < 0 || beta < 0){
     elog(ERROR, "BETA: Can't have negative alpha/beta parameters");
     PG_RETURN_NULL();
  } else {
     boost::random::beta_distribution<double> dist(alpha, beta);
     return _gen_rand(fcinfo, dist, 2);
  }
}

RAND_FUNCTION(gen_binomial) {
  int t = get_param<int>(fcinfo, 0);
  float p  = get_param<double>(fcinfo, 1);
  if (t < 0 || p < 0 || p > 1){
     elog(ERROR, "BINOMIAL: Input parameters outside of range");
     PG_RETURN_NULL();
  } else {
     boost::random::binomial_distribution<int> dist(t,p);
     return _gen_rand(fcinfo, dist, 2);
  }
}

