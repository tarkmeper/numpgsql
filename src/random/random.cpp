extern "C"  {
        #include "postgres.h"
        #include "catalog/pg_type.h"
        #include "fmgr.h"
        #include "math.h"
}
#include <string>
#include <iostream>
#include <sstream>
#include "random.h"
#include "../array.h"

std::mt19937 rgen;

ArrDims extract_shape(ArrayType* arr) {
  static constexpr int LBS_VALS[] = { 1, 1, 1, 1, 1, 1 };
  ArrDims res;
  res.dims = (int*)(ARR_DATA_PTR(arr));
  res.ndims = ARR_DIMS(arr)[0];
  res.n = res.dims[0];
  memcpy(res.lbs, LBS_VALS, sizeof(int) * MAXDIM);
  for ( int i = 1; i < res.ndims; ++i) { res.n *= res.dims[i]; }
  return res;
}

extern "C" { PG_FUNCTION_INFO_V1(set_seed); Datum set_seed(PG_FUNCTION_ARGS); }
 Datum set_seed(PG_FUNCTION_ARGS) {
  const int s = PG_GETARG_INT32(0);
  rgen.seed(s);
  PG_RETURN_NULL();
}


extern "C" { PG_FUNCTION_INFO_V1(get_rgen_state); Datum get_rgen_state(PG_FUNCTION_ARGS); }
Datum get_rgen_state(PG_FUNCTION_ARGS) {
  std::stringstream ss;
  ss << rgen;

  const std::string val = ss.str();

  //Now move over
  text *new_t = (text *) palloc(val.size() + VARHDRSZ);
  SET_VARSIZE(new_t, val.size() + VARHDRSZ);
  memcpy(VARDATA(new_t), val.c_str(), val.size());

  PG_RETURN_TEXT_P(new_t);
}


extern "C" { PG_FUNCTION_INFO_V1(set_rgen_state); Datum set_rgen_state(PG_FUNCTION_ARGS); }
Datum set_rgen_state(PG_FUNCTION_ARGS) {
  text  *arg1 = PG_GETARG_TEXT_PP(0);  std::string val(VARDATA(arg1));
  std::istringstream ss(val);
  ss >> rgen;

  PG_RETURN_NULL();
}
