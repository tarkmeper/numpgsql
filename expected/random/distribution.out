SELECT random_seed(1);
 random_seed 
-------------
 
(1 row)

SELECT random_beta(2,3, '{2,3}'::integer[]);
                                                    random_beta                                                    
-------------------------------------------------------------------------------------------------------------------
 {{0.175285250340353,0.197240925500585,0.881228480984027},{0.495397936679039,0.121173507019459,0.249182047453486}}
(1 row)

SELECT random_beta(-1,3, '{2,3}'::integer[]);
ERROR:  BETA: Can't have negative alpha/beta parameters
SELECT random_beta(2,-3, '{2,3}'::integer[]);
ERROR:  BETA: Can't have negative alpha/beta parameters
SELECT random_binomial(2,0.5, '{2,3}'::integer[]);
  random_binomial  
-------------------
 {{2,2,1},{2,2,1}}
(1 row)

SELECT random_binomial(-2,0.5, '{2,3}'::integer[]);
ERROR:  BINOMIAL: Input parameters outside of range
SELECT random_binomial(2,3, '{2,3}'::integer[]);
ERROR:  BINOMIAL: Input parameters outside of range
SELECT random_binomial(2,-3, '{2,3}'::integer[]);
ERROR:  BINOMIAL: Input parameters outside of range
